### ISSUE TITLE
GDB Error with Run DUB

### ISSUE THREAD

**Issue by abaga129** [intellij-dlanguage/intellij-dlanguage#526]
I'm receiving the following error with Intellij Community on Linux

Intellij IDEA CE 2019.2
GDB Version: 8.3.1

This is a fresh project created using the plugin.  I only added two lines of code to the default `app.d`
```
GDB error

java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
	at uk.co.cwspencer.gdb.gdbmi.GdbMiParser2.parseBreakpointHitLineFrameLine(GdbMiParser2.java:64)
	at uk.co.cwspencer.gdb.gdbmi.GdbMiParser2.parseBreakpointHitLine(GdbMiParser2.java:1203)
	at uk.co.cwspencer.gdb.gdbmi.GdbMiParser2.parseExecLine(GdbMiParser2.java:666)
	at uk.co.cwspencer.gdb.gdbmi.GdbMiParser2.parseLine(GdbMiParser2.java:581)
	at uk.co.cwspencer.gdb.gdbmi.GdbMiParser2.process(GdbMiParser2.java:479)
	at uk.co.cwspencer.gdb.Gdb.runGdb(Gdb.java:310)
	at uk.co.cwspencer.gdb.Gdb.access$000(Gdb.java:43)
	at uk.co.cwspencer.gdb.Gdb$1.run(Gdb.java:88)
	at java.base/java.lang.Thread.run(Thread.java:830)
```
GDB Console output
```
08:15:30.395 0> /usr/bin/gdb --interpreter=mi2
08:15:30.406 1> -file-exec-and-symbols /home/ethanpreker/Programming/DAppTest/dapptest
08:15:30.415 2> -break-insert -f /home/ethanpreker/Programming/DAppTest/source/app.d:6
08:15:30.500 3> -list-features
08:15:30.500 4> -exec-run
08:15:30.598 < [notify] thread-group-added: [id: "i1"]
GNU gdb (GDB) 8.3.1
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later http://gnu.org/licenses/gpl.html
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
http://www.gnu.org/software/gdb/bugs/.
Find the GDB manual and other documentation resources online at:
    http://www.gnu.org/software/gdb/documentation/.

For help, type "help".
Type "apropos word" to search for commands related to "word".
08:15:30.643 <1 [immediate] done
08:15:30.646 <2 [immediate] done: [bkpt: {number: "1", type: "breakpoint", disp: "keep", enabled: "y", addr: "0x0000000000053824", func: "D main", file: "source/app.d", fullname: "/home/ethanpreker/Programming/DAppTest/source/app.d", line: "6", thread-groups: ["i1"], times: "0", original-location: "/home/ethanpreker/Programming/DAppTest/source/app.d:6"}]
08:15:30.655 <3 [immediate] done: [features: ["frozen-varobjs", "pending-breakpoints", "thread-info", "data-read-memory-bytes", "breakpoint-notifications", "ada-task-info", "language-option", "info-gdb-mi-command", "undefined-command-error-code", "exec-run-start-option", "data-disassemble-a-option", "python"]]
08:15:30.655 <3 [notify] thread-group-started: [id: "i1", pid: "20453"]
08:15:30.655 < [notify] thread-created: [id: "1", group-id: "i1"]
08:15:30.680 < [notify] breakpoint-modified: [bkpt: {number: "1", type: "breakpoint", disp: "keep", enabled: "y", addr: "0x00005555555a7824", func: "D main", file: "source/app.d", fullname: "/home/ethanpreker/Programming/DAppTest/source/app.d", line: "6", thread-groups: ["i1"], times: "0", original-location: "/home/ethanpreker/Programming/DAppTest/source/app.d:6"}]
08:15:30.680 < [notify] library-loaded: [id: "/lib64/ld-linux-x86-64.so.2", target-name: "/lib64/ld-linux-x86-64.so.2", host-name: "/lib64/ld-linux-x86-64.so.2", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7fd4100", to: "0x00007ffff7ff2c04"]
08:15:30.680 <4 [immediate] running
08:15:30.680 <4 [exec] running: [thread-id: "all"]
08:15:30.681 < [notify] library-loaded: [id: "/usr/lib/libpthread.so.0", target-name: "/usr/lib/libpthread.so.0", host-name: "/usr/lib/libpthread.so.0", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7f6cb30", to: "0x00007ffff7f7b995"]
08:15:30.681 < [notify] library-loaded: [id: "/usr/lib/libm.so.6", target-name: "/usr/lib/libm.so.6", host-name: "/usr/lib/libm.so.6", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7e2e370", to: "0x00007ffff7ec9cb8"]
08:15:30.681 < [notify] library-loaded: [id: "/usr/lib/librt.so.1", target-name: "/usr/lib/librt.so.1", host-name: "/usr/lib/librt.so.1", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7e176f0", to: "0x00007ffff7e1a950"]
08:15:30.681 < [notify] library-loaded: [id: "/usr/lib/libdl.so.2", target-name: "/usr/lib/libdl.so.2", host-name: "/usr/lib/libdl.so.2", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7e10210", to: "0x00007ffff7e10fe9"]
08:15:30.681 < [notify] library-loaded: [id: "/usr/lib/libgcc_s.so.1", target-name: "/usr/lib/libgcc_s.so.1", host-name: "/usr/lib/libgcc_s.so.1", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7df8020", to: "0x00007ffff7e08575"]
08:15:30.681 < [notify] library-loaded: [id: "/usr/lib/libc.so.6", target-name: "/usr/lib/libc.so.6", host-name: "/usr/lib/libc.so.6", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7c53670", to: "0x00007ffff7d9e33f"]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
08:15:30.730 < [notify] breakpoint-modified: [bkpt: {number: "1", type: "breakpoint", disp: "keep", enabled: "y", addr: "0x00005555555a7824", func: "D main", file: "source/app.d", fullname: "/home/ethanpreker/Programming/DAppTest/source/app.d", line: "6", thread-groups: ["i1"], times: "1", original-location: "/home/ethanpreker/Programming/DAppTest/source/app.d:6"}]
```

Any ideas on what could cause this?

**Comment by SingingBush** [comment]
Seems to break in same line as #427 

https://github.com/intellij-dlanguage/intellij-dlanguage/blob/develop/src/main/java/uk/co/cwspencer/gdb/gdbmi/GdbMiParser2.java#L64

It should've really been dealt with then. I'm not particularly happy with that _uk.co.cwspencer.gdb_ package in general. Ideally it should be maintained separately and pulled in as a dependency.

**Comment by abaga129** [comment]
Thanks for looking at this so quickly.  It may be a few days before I can really dig into the code and try to fix this.  However, I did just pull the change you made 4f1ab0a and bulilt/ran it.  The debugger is getting further it seems.  Here is the log from the GDB Console

```bash
23:46:38.616 0> /usr/bin/gdb --interpreter=mi2
23:46:38.620 1> -file-exec-and-symbols /home/reker/IdeaProjects/DAppTest/dapptest
23:46:38.620 2> -break-insert -f /home/reker/IdeaProjects/DAppTest/source/app.d:6
23:46:38.658 3> -list-features
23:46:38.658 4> -exec-run
23:46:38.763 < [notify] thread-group-added: [id: "i1"]
GNU gdb (GDB) 8.3
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later http://gnu.org/licenses/gpl.html
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
http://www.gnu.org/software/gdb/bugs/.
Find the GDB manual and other documentation resources online at:
    http://www.gnu.org/software/gdb/documentation/.

For help, type "help".
Type "apropos word" to search for commands related to "word".
23:46:38.796 <1 [immediate] done
23:46:38.796 <2 [immediate] done: [bkpt: {number: "1", type: "breakpoint", disp: "keep", enabled: "y", addr: "0x0000000000054824", func: "D main", file: "source/app.d", fullname: "/home/reker/IdeaProjects/DAppTest/source/app.d", line: "6", thread-groups: ["i1"], times: "0", original-location: "/home/reker/IdeaProjects/DAppTest/source/app.d:6"}]
23:46:38.797 <3 [immediate] done: [features: ["frozen-varobjs", "pending-breakpoints", "thread-info", "data-read-memory-bytes", "breakpoint-notifications", "ada-task-info", "language-option", "info-gdb-mi-command", "undefined-command-error-code", "exec-run-start-option", "data-disassemble-a-option", "python"]]
23:46:38.797 <3 [notify] thread-group-started: [id: "i1", pid: "7897"]
23:46:38.797 < [notify] thread-created: [id: "1", group-id: "i1"]
23:46:38.797 < [notify] breakpoint-modified: [bkpt: {number: "1", type: "breakpoint", disp: "keep", enabled: "y", addr: "0x00005555555a8824", func: "D main", file: "source/app.d", fullname: "/home/reker/IdeaProjects/DAppTest/source/app.d", line: "6", thread-groups: ["i1"], times: "0", original-location: "/home/reker/IdeaProjects/DAppTest/source/app.d:6"}]
23:46:38.797 < [notify] library-loaded: [id: "/lib64/ld-linux-x86-64.so.2", target-name: "/lib64/ld-linux-x86-64.so.2", host-name: "/lib64/ld-linux-x86-64.so.2", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7fd4100", to: "0x00007ffff7ff2c64"]
23:46:38.797 <4 [immediate] running
23:46:38.797 <4 [exec] running: [thread-id: "all"]
23:46:38.797 < [notify] library-loaded: [id: "/usr/lib/libpthread.so.0", target-name: "/usr/lib/libpthread.so.0", host-name: "/usr/lib/libpthread.so.0", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7f69b20", to: "0x00007ffff7f77d85"]
23:46:38.797 < [notify] library-loaded: [id: "/usr/lib/libm.so.6", target-name: "/usr/lib/libm.so.6", host-name: "/usr/lib/libm.so.6", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7e2b370", to: "0x00007ffff7ec6cb8"]
23:46:38.797 < [notify] library-loaded: [id: "/usr/lib/librt.so.1", target-name: "/usr/lib/librt.so.1", host-name: "/usr/lib/librt.so.1", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7e146f0", to: "0x00007ffff7e17950"]
23:46:38.797 < [notify] library-loaded: [id: "/usr/lib/libdl.so.2", target-name: "/usr/lib/libdl.so.2", host-name: "/usr/lib/libdl.so.2", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7e0d210", to: "0x00007ffff7e0dfe9"]
23:46:38.797 < [notify] library-loaded: [id: "/usr/lib/libgcc_s.so.1", target-name: "/usr/lib/libgcc_s.so.1", host-name: "/usr/lib/libgcc_s.so.1", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7df5020", to: "0x00007ffff7e05525"]
23:46:38.834 < [notify] library-loaded: [id: "/usr/lib/libc.so.6", target-name: "/usr/lib/libc.so.6", host-name: "/usr/lib/libc.so.6", symbols-loaded: "0", thread-group: "i1", from: "0x00007ffff7c54630", to: "0x00007ffff7d9c89f"]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
23:46:38.859 < [notify] breakpoint-modified: [bkpt: {number: "1", type: "breakpoint", disp: "keep", enabled: "y", addr: "0x00005555555a8824", func: "D main", file: "source/app.d", fullname: "/home/reker/IdeaProjects/DAppTest/source/app.d", line: "6", thread-groups: ["i1"], times: "1", original-location: "/home/reker/IdeaProjects/DAppTest/source/app.d:6"}]

Breakpoint 1, D main () at source/app.d:6
6        writeln("Edit source/app.d to start your project.");
23:46:38.859 < [exec] stopped: [reason: "breakpoint-hit", disp: "keep", bkptno: "1", null, thread-id: "1", stopped-threads: "all", core: "11"]
```
There's no error now, but the debugger in IntelliJ doesn't know that the breakpoint-hit has happened. 
![Debugger](https://imgur.com/x1ZcmkK.png)

**Comment by SingingBush** [comment]
Yeah, the change I made is far from complete. I just looked at it briefly before going to sleep. The intention was just to make it more obvious that _parseBreakpointHitLineFrameLine_ could return null. I've also done a few obvious tweaks to _GdbMiMessageConverter_ now in a similar manner but as you've mentioned that section that `List<GdbMiResult> parseFrameLine(final String line)` method needs some work. Using a regex to parse the text seems less than ideal. I think the code that parses whatever is coming out of the debugger should be in it's own method (perhaps in _GdbUtil_) so that it can have decent test coverage.
